api/src/handlers.rs

```diff
@@ -1,5 +1,237 @@
+}
+
+pub async fn add_ai(info: Path<(String, String, String, String)>) -> impl Responder {
+    let info = info.into_inner();
+    let ai = CribbageAi {
+        name: info.0,
+        by: info.1,
+        description: info.2,
+        uri: info.3,
+    };
+
+    let connection_info = get_connection_info!();
+
+    let client = cosmos_connection!(connection_info);
+
+    let client = client.into_database_client(&connection_info.database_name);
+    let client = client.into_collection_client(&connection_info.collection_name);
+
+    // let response = client.list_documents().execute::<CribbageAi>().await;
+    // if response.is_err() {
+    //     return HttpResponse::NotFound().body(format!(
+    //         "No Collection installed for db: {}",
+    //         connection_info.database_name
+    //     ));
+    // }
+
+    // let mut found = false;
+    // for docs in response.unwrap().documents {
+    //     if docs.id == connection_info.collection_name {
+    //         found = true;
+    //         break;
+    //     }
+    // }
+
+    // if found == false {
+    //     return HttpResponse::NotFound().body(format!(
+    //         "Collection {} not found.  Configure Cosmos!",
+    //         connection_info.collection_name
+    //     ));
+    // }
+
+    // let client = client.with_collection_client(connection_info.collection_name);
+
+    let response = client
+        .list_documents()
+        .with_max_item_count(100)
+        .execute::<CribbageAi>()
+        .await;
+
+    //
+    //  check for duplicate names
+    //
+    if response.is_ok() {
+        for doc in response.unwrap().documents {
+            if doc.document.name == ai.name {
+                return HttpResponse::Ok().body(format!("found duplicate name: {}\n.  Remove it then add again if you want to update it.\nDoing nothing.", ai.name));
+            };
+        }
+    }
+
+    let mut partition_keys = PartitionKeys::default();
+    match partition_keys.push(&ai.name) {
+        Ok(_) => {}
+        Err(e) => {
+            return HttpResponse::Ok()
+                .body(format!("Error creating partition key: {}", e.to_string()));
+        }
+    }
+
+    // let response = Some(
+    //     client
+    //         .create_document(Context::new(), &ai, CreateDocumentOptions::new())
+    //         .await,
+    // );
+
+    match client
+        .create_document()
+        .with_partition_keys(&partition_keys)
+        .execute_with_document(&ai)
+        .await
+    {
+        Ok(_) => {
+            return HttpResponse::Ok().body(serde_json::to_string(&ai).unwrap());
+        }
+        Err(e) => {
+            return HttpResponse::Ok().body(format!("Error adding document: {}", e.to_string()));
+        }
+    };
+}
+
+pub async fn test_cosmos() -> impl Responder {
+    let connection_info = get_connection_info!();
+    let client = cosmos_connection!(connection_info);
+    let client = client.into_database_client(connection_info.database_name);
```

Comment:
Can you chain these calls for readability, instead of the multiple `let client = client.` lines?